### ISSUE TITLE
Roundcube can't connect to Cyrus with SSL

### ISSUE THREAD

**Issue by majsternc** [roundcube/roundcubemail#6877]
After upgrade Debian 9 to Debian 10 on which Roundcube is installed, I can't connect to Cyrus.
Cyrus is installed on a different/separate server.

In logs of a Cyrus I can see:
Jul 30 17:26:26 hostname cyrus/imaps[1605]: imaps TLS negotiation failed: [192.168.3.217]
Jul 30 17:26:26 hostname cyrus/imaps[1605]: Fatal error: tls_start_servertls() failed
Jul 30 17:26:26 hostname cyrus/master[30285]: process 1605 exited, status 75
Jul 30 17:26:26 hostname cyrus/master[30285]: service imaps pid 1605 in BUSY state: terminated abnormally

Same config on Debian 9 works like a charm.

What can be an issue with that?

**Comment by alecpl** [comment]
Probably something with ssl setup on the system or in php or cyrus. Try setting some connection options, e.g.
```
$config['imap_conn_options'] = array(
  'ssl' => array(
     'crypto_method' => STREAM_CRYPTO_METHOD_TLSv1_2_CLIENT
   )
);
```
Also, switching ssl:// to tls:// or vice-versa may help.

**Comment by majsternc** [comment]
Your suggestion didn't work so far. I've tried with another crypto_method as well. 

After enabling roudncube debug mode I've got these errors:
Warning: stream_socket_client(): SSL operation failed with code 1. OpenSSL Error messages: error:141A318A:SSL routines:tls_process_ske_dhe:dh key too small in /usr/share/roundcube/program/lib/Roundcube/rcube_imap_generic.php on line 959

Warning: stream_socket_client(): Failed to enable crypto in /usr/share/roundcube/program/lib/Roundcube/rcube_imap_generic.php on line 959

Warning: stream_socket_client(): unable to connect to ssl://mail.hostname.pl:993 (Unknown error) in /usr/share/roundcube/program/lib/Roundcube/rcube_imap_generic.php on line 959
IMAP Error in /usr/share/roundcube/program/lib/Roundcube/rcube_imap.php (196): Login failed for majster from 192.168.3.118. Could not connect to ssl://mail.hostname.pl:993: Unknown reason

**Comment by alecpl** [comment]
That OpenSSL error message suggests it's not something we can fix on Roundcube side. Some OpenSSL policy prevents use of that ssl key.

**Comment by majsternc** [comment]
Ok, I've done it.

Sollution:
I suggest to remove
CipherString = DEFAULT@SECLEVEL=2
from /etc/ssl/openssl.cnf.
source: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=907788

So you have to comment out this line
CipherString = DEFAULT@SECLEVEL=2
from file /etc/ssl/openssl.cnf on roundcube server side.

Explanation:
https://wiki.debian.org/ContinuousIntegration/TriagingTips/openssl-1.1.1

**Comment by gr4y** [comment]
Hey there,

I ran into the same problem (with dovecot instead of cyrus).
And I tried your solution, bit it didn't work for me, so I researched further and discovered https://weakdh.org/sysadmin.html, which led me in the right direction. 

So my solution was to generate a 4096 bit long DHParam:
`openssl dhparam -out /etc/ssl/certs/dhparams.pem 4096` 

And I had to add this to `/etc/dovecot/dovecot.conf`
`ssl_dh = </etc/ssl/certs/dhparams.pem` 

Everything is running smoothly now!
Hope this helps you!
